File: compareDataSet.java
import java.util.*;

public void compareDataSet(Vector<Vector<String>> leftData, Vector<Vector<String>> rightData, Long ExcelCounter)
{
    Map<String, Vector<String>> leftMap = new HashMap<>();
    Map<String, Vector<String>> rightMap = new HashMap<>();

    for(Vector<String> v: leftData) {
        leftMap.put(v.get(7), v);
    }

    for(Vector<String> v: rightData) {
        rightMap.put(v.get(7), v);
    }

    Set<String> keys = new TreeSet<>(leftData.keySet());
    keys.retainAll(rightData.keySet());

    Vector<Integer> cells = new Vector<>();

    for(String key: keys) {

        int j = 3;
        Iterator<String> lefti = leftMap.get(key).iterator();
        Iterator<String> righti = rightMap.get(key).iterator();

        while (lefti.hasNext() && righti.hasNext()) {
            String ls = lefti.next();
            String rs = righti.next();
            if (ls.compareTo(rs) != 0) {
                cells.add(j);
            }
            j++;
        }

        while (lefti.hasNext()) {
            cells.add(j);
            lefti.next();
            j++;
        }

        while (righti.hasNext()) {
            cells.add(j);
            righti.next();
            j++;
        }

        // Add rows 0, 1, 2

        if (cells.size() > 0) {
            if (additional_info == 0 || additional_info == 2) {
                exg.push(leftMap.get(key), ExcelCounter);
                exg.color(ExcelCounter, cells);
                ExcelCounter++;
            }
            if (additional_info == 1 || additional_info == 2) {
                exg.push(rightMap.get(key), ExcelCounter);
                exg.color(ExcelCounter, cells);
                ExcelCounter++;
            }
        } else if (discrepancy == 1) {
            if (additional_info == 0 || additional_info == 2) {
                exg.push(leftMap.get(key), ExcelCounter);
                ExcelCounter++;
            }
            if (additional_info == 1 || additional_info == 2) {
                exg.push(rightMap.get(key), ExcelCounter);
                ExcelCounter++;
            }
        }

        cells.clear();
    }

    leftMap.removeAll(keys);
    rightMap.removeAll(keys);

    if(additional_info == 0 || additional_info == 2) {
        for(String key: leftMap.keySet()) {
            for (int j = 0; j < leftMap.get(key).size(); j++) {
                cells.add(j + 3);
            }
            exg.push(leftMap.get(key), ExcelCounter);
            exg.color(ExcelCounter, cells);
            ExcelCounter++;

            cells.clear();
        }
    }

    if(additional_info == 0 || additional_info == 2) {
        for (String key: rightMap.keys()) {
            for (int j = 0; j < rightMap.get(key).size(); j++) {
                cells.add(j + 3);
            }
            exg.push(rightMap.get(key), ExcelCounter);
            exg.color(ExcelCounter, cells);
            ExcelCounter++;

            cells.clear();
        }
    }

}

File: compareDataSet.py
import operator


class ExcelGenerator:

    def push(self, alist, counter):
        print(alist, end=' -> ')

    def color(self, counter, cells):
        print(cells)


def compareDataSet(leftData, rightData, ExcelCounter):

    leftData.sort(key=operator.itemgetter(1))
    rightData.sort(key=operator.itemgetter(1))

    leftMap = {}
    rightMap = {}

    for item in leftData:
        leftMap[item[1]] = item
    for item in rightData:
        rightMap[item[1]] = item

    cells = list()

    # Remove
    additional_info = 2
    exg = ExcelGenerator()

    keys = set(leftMap.keys())
    keys = keys.intersection(rightMap.keys())

    for key in keys:

        lefti = list(leftMap[key])
        righti = list(rightMap[key])

        j = 0

        while len(lefti) > 0 and len(righti) > 0:
            ls = lefti.pop(0)
            rs = righti.pop(0)
            if (ls != rs):
                cells.append(j)
            j += 1

        cells += [i for i in range(j, j + len(lefti))]
        cells += [i for i in range(j, j + len(righti))]

        del lefti
        del righti

        if len(cells) > 0:
            if additional_info == 0 or additional_info == 2:
                exg.push(leftMap[key], ExcelCounter)
                exg.color(ExcelCounter, cells)
                ExcelCounter += 1
            if additional_info == 1 or additional_info == 2:
                exg.push(rightMap[key], ExcelCounter)
                exg.color(ExcelCounter, cells)
                ExcelCounter += 1
        elif False:
            if additional_info == 0 or additional_info == 2:
                exg.push(leftMap[key], ExcelCounter)
                ExcelCounter += 1
            if additional_info == 1 or additional_info == 2:
                exg.push(rightMap[key], ExcelCounter)
                ExcelCounter += 1

        del leftMap[key]
        del rightMap[key]

        cells.clear()

    if additional_info == 0 or additional_info == 2:
        for key in leftMap.keys():
            cells += [j + 3 for j in range(len(leftMap[key]))]
            exg.push(leftMap[key], ExcelCounter)
            exg.color(ExcelCounter, cells)
            ExcelCounter += 1
            cells.clear()

    if additional_info == 1 or additional_info == 2:
        for key in rightMap.keys():
            cells += [j + 3 for j in range(len(rightMap[key]))]
            exg.push(rightMap[key], ExcelCounter)
            exg.color(ExcelCounter, cells)
            ExcelCounter += 1
            cells.clear()


leftData = [
    ["ABC", 213, 431],
    ["ABC", 132, 435],
    ["ABC", 123, 432],
    ["RAJI", 333, 999]
]
rightData = [
    ["ABC", 132, 437],
    ["PQR", 123, 432],
    ["XYZ", 213, 438],
    ["RAGHU", 111, 999]
]

compareDataSet(leftData, rightData, 0)
